### PR TITLE
ci: circle CI voter and release jobs

### PR DESCRIPTION
**Pull request by bd82**

### CODE REVIEW

**Comment by bd82** [comment]
resolves #22 

@matthew-dean: 

### Build failures
There seems to be a cyclic dependency, which needs to be resolved for the build to succeed.
-   
image


### Publishing
- You will need to insert a secure env variable into circleci called `NPM_TOKEN` to enable publishing to npmjs.com from circleci.
  - https://docs.npmjs.com/creating-and-viewing-access-tokens

I recommend we merge this PR and fix the project structure in a different PR.

**Comment by matthew-dean** [comment]
> You will need to insert a secure env variable into circleci called NPM_TOKEN to enable publishing to npmjs.com from circleci

Do I want to do that? 😬  I've been doing it locally with `lerna publish` and then I get the benefit of choosing a version. I'm not sure I want to automate publishing; that seems too scary. Can you talk through how that works in practice?

**Comment by matthew-dean** [comment]
Re: the cyclic dependency - Yes, I built a Rollup plugin for Jess, but ended up using it in the Jess command-line to create JS bundles. I'm not sure how to resolve that. I want to expose both libraries to NPM. Both need each other. And for me, that's only caused build warnings and not errors, I thought. 🤔

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      - run: echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" >> ~/.npmrc
+      # To help debug when lerna detects changes to the working tree and fails the publish
+      - run: git status
+      - run: yarn run release:publish
+
+workflows:
+  version: 2
+  voter:
+    jobs:
+      - build-node12
```
Comment (matthew-dean) [inline .circleci/config.yml:47]: You're sure Node 10 isn't needed? I'm still personally on that version for work stuff. 🤷‍♂️

**Comment by bd82** [comment]
> Do I want to do that? 😬 I've been doing it locally with lerna publish and then I get the benefit of choosing a version. I'm not sure I want to automate publishing; that seems too scary. Can you talk through how that works in practice?

The version choosing still happens manually by the user (see CONTRIBUTING.md release section).
Only the publishing happens on circleci which is safer and more consistent than your local dev env.

File: .circleci/config.yml (hunk 1 above)
Comment (bd82) [reply .circleci/config.yml:47]: We can always add it later, but node10 will reach EOL in ~2 months, so maybe it is not worth it

**Comment by bd82** [comment]
> Re: the cyclic dependency - Yes, I built a Rollup plugin for Jess, but ended up using it in the Jess command-line to create JS bundles. I'm not sure how to resolve that. I want to expose both libraries to NPM. Both need each other. And for me, that's only caused build warnings and not errors, I thought. 🤔

It is tricky, I once had an issue like that with a programmatic API (not CLI) and had to inject some implementations using DI to avoid a direct dependency. Not sure what to do about your CLI use case.

The reason it works for you locally may be because it is using a **previously** built version of the rollup plugin to bundle your resources but in the CI env there is no such previous version.

Basically these cyclic deps often create issues when there are build steps involved in the project.
You could perhaps try to workaround the problem by running a compilation step first project wide and only then running the 
CI build (minus compilation).

But it is best if you can figure out a way to avoid this cycle, as issues with it will likely keep popping up in different scenarios (that was my personal experience)

File: .circleci/config.yml (hunk 1 above)
Comment (matthew-dean) [reply .circleci/config.yml:47]: Sounds fair.

**Comment by matthew-dean** [comment]
> The version choosing still happens manually by the user 

Okay, that makes sense.

> But it is best if you can figure out a way to avoid this cycle, as issues with it will likely keep popping up in different scenarios (that was my personal experience)

I'll see what I can do. Thanks!

**Comment by matthew-dean** [comment]
@bd82 If I can resolve this cyclic dependency, would I just run `lerna ci` locally to verify it's working?

**Comment by bd82** [comment]
> If I can resolve this cyclic dependency, would I just run lerna ci locally to verify it's working?

I think you would run:
- `lerna clean`
- ensure root `node_modules` was also cleaned (if not `rm -rf node_modules`)
- yarn ci

`yarn ci` will run the top level `ci` script which will run `lerna ci` and also some other root level validations (currently linting)

**Comment by matthew-dean** [comment]
Closing as obsolete: this 2021 CircleCI setup predates and is superseded by the current GitHub Actions gates (`.github/workflows/pr-quality-gate.yml` and `less-alpha-readiness.yml`). The included package.json change is also long stale. If CircleCI is ever wanted again it should be a fresh PR against the current tree.
